module/kops.tf

```diff
@@ -4,19 +4,58 @@ resource "null_resource" "check_kops_version" {
   }
 }
 
+data "template_file" "tmpl_values" {
+  template = "${file("${path.module}/values.yaml.tmpl")}"
+
+  vars {
+    cluster_fqdn          = "${local.cluster_fqdn}"
+    kubernetes_version    = "${var.kubernetes_version}"
+    kops_dns_mode         = "${var.kops_dns_mode}"
+    kops_s3_bucket_id     = "${var.kops_s3_bucket_id}"
+    kubernetes_networking = "${var.kubernetes_networking}"
+    zones                 = "${join(",", data.aws_availability_zones.available.names)}"
+    node_asg_desired      = "${var.node_asg_desired}"
+    master_azs            = "${local.master_azs}"
+    vpc_id                = "${var.vpc_id}"
+    ami_name              = "${local.ami_name}"
+    k8s_apiserver_options = "${join("\n  ", compact(local.k8s_apiserver_options))}"
+  }
+}
+
+resource "null_resource" "generate_values" {
+  triggers {
+    template = "${data.template_file.tmpl_values.rendered}"
+  }
+
+  provisioner "local-exec" {
+    command = "echo \"${data.template_file.tmpl_values.rendered}\" > values.yaml"
+  }
+}
+
+resource "null_resource" "generate_template" {
+  depends_on = ["null_resource.generate_values"]
+
+  provisioner "local-exec" {
+    command = "kops toolbox template --values values.yaml --template ${path.module}/cluster_config.yaml.tmpl --output cluster_config.yaml"
+  }
+}
+
 resource "null_resource" "create_cluster" {
-  depends_on = [ "null_resource.check_kops_version" ]
+  depends_on = ["null_resource.check_kops_version", "null_resource.generate_template"]
+
   provisioner "local-exec" {
-    command = "kops create cluster --cloud=aws --dns ${var.kops_dns_mode} --authorization RBAC --networking ${var.kubernetes_networking} --zones=${join(",", data.aws_availability_zones.available.names)} --node-count=${var.node_asg_desired} --master-zones=${local.master_azs} --target=terraform --api-loadbalancer-type=public --vpc=${var.vpc_id} --state=s3://${var.kops_s3_bucket_id} --kubernetes-version ${var.kubernetes_version} ${local.cluster_fqdn}"
+    command = "kops create -f cluster_config.yaml --state=s3://${var.kops_s3_bucket_id} && kops create secret --state=s3://vend-test-kops-sydney --name cluster.ap-southeast-2.test.vendhq.com sshpublickey admin -i ~/.ssh/id_rsa.pub && kops update cluster ${var.cluster_fqdn} --state=s3://${var.kops_s3_bucket_id} --target=terraform --yes"
```

Comment:
Will this make every cluster use the templates, regardless of whether those features are wanted?

Reply:
Also here you have a hard coded s3 bucket name `vend-test-kops-sydney`

Reply:
Is the secret required for every cluster? Even if we want just a vanilla kops cluster without these extra features? If not, then I would move this command to separate `null_resource`.